### ISSUE TITLE
New AFHDS 2A module/receiver failsafe.

### ISSUE THREAD

**Issue by gunsmithfiamma** [MikeBland/mbtx#33]
Hello Mike,
since Flysky/Turnigy finally released a new TX module for AFHDS 2A protocol, is there a possibility to add a Failsafe menu in the firmware? 
As you already know, for AFHDS 2A receivers, any failsafe function can be programmed only from the radio.
Er9x works great with the new 9X and AFHDS 2A module , but without failsafe can be dangerous.
Thanks a lot for your hard work.
gunsmithfiamma

**Comment by MikeBland** [comment]
Currently, I have no idea if the AFHDS2A Tx module can send failsafe values.
Since we send the channel data using PPM to the module, I'm not sure we have any way of telling the module what the failsafe values are.
If you have a manual for the module (I don't have such a module), and it details how failsafe may be done, please let me know and I can look inot whether it is possible to add to er9x.

**Comment by gunsmithfiamma** [comment]
Sorry for the late answer.
Yes the module can send failsafe values, because in the new firmware of the radio (the one supplied with the AFHDS 2A module) there is a failsafe menu and it works. Since the new firmware is not released at the moment, I used a spare main board for Er9X, so I still have the original board with the original firmware on. I can try to extract the firmware from the board if it will help you.
The set up is very simple: Put the stick/switches/potentiometer in the position you need, go in the failsafe menu and turn on the function on the channel you need the failsafe. ![img_20171206_215656_risultato](https://user-images.githubusercontent.com/20949209/33685239-4b424c9a-dad1-11e7-9f6a-416aa7bfd052.jpg)

**Comment by MikeBland** [comment]
Yes, please try to extract the firmware from the original main board.

Mike.

**Comment by gunsmithfiamma** [comment]
Ok, I managed to extract the hex file from the original board.
I hope that it can be useful. 
Cheers 
Francesco
[TGY9xAFhds2a.hex.zip](https://github.com/MikeBland/mbtx/files/1541802/TGY9xAFhds2a.hex.zip)
P.s.: I'm relatively new to this kind of procedure so, if there is some error, don't get mad :-)

**Comment by MikeBland** [comment]
I've got that OK, but I think I need the stock EEPROM contents as well if it is possible for you to extract that and post it as well.
I have flashed this firmware, and an EEPROM from an original board, and have this running, although it is not easy as my 9X has the telemetry mod so I get the "SWITCH ERROR" at power on and have to short some pins on the programming header to clear it.
I tried using the falisafe menu, but every time I enter the menu, everything says "NOR", even if I change to "F/S and set a value.
I note that the stock firmware has an option of PPM or PCM. I suspect that it may be necessary to use the PCM to the module to get the failsafe function. If this is the case, then it will be necessary to work out what the PCM coding is, and try to add that to er9x, but, if you have a M64 processor, I doubt the extra code will fit.
Even if possible, this won't be a change that can be done quickly or easily.

**Comment by gunsmithfiamma** [comment]
Since I'm a Mac user, is there a specific application to download the EEPROM from the board?
I'm sorry if this is a silly question, but I'm an amateur in electronic, and this is a new territory for me :-) (I'm a professional Gunsmith so, open source radio systems, aren't really my craft)

**Comment by lexrupy** [comment]
Hello, any progress on that?
Also I tried to build the er9x firmware on linux, it built notmally, but after a flashed it out the radio doesn´t boot. So I flasshed the last you provided on openrcforuns test page er9xProv822m.zip.
Also my compiled binary does not have the same size in bytes of yours, even with many combinations of compiler directives.

**Comment by macem** [comment]
it is possible to support this module ? Regards  Marcin
